packages/node/src/exec/win32.ts

```diff
@@ -1,7 +1,117 @@
+export const createWin32Process: CreateOSProcess = function*(command, options) {
+  let stdin = new Channel<string>();
+  let stdout = new Channel<string>();
+  let stderr = new Channel<string>();
+  let tail: string[] = [];
+
+  let getResult = Deferred<Result>();
+
+  function addToTail(chunk: string) {
+    if (tail.length < 100) {
+      tail.push(chunk);
+    }
+  }
+
+  function* join(): Operation<ExitStatus> {
+    let result: Result = yield getResult.promise;
+    if (result.type === "status") {
+      let [code, signal] = result.value;
+      return { command, options, code, signal, tail };
+    } else {
+      throw result.value;
+    }
+  }
+
+  function* expect(): Operation<ExitStatus> {
+    let status: ExitStatus = yield join();
+    if (status.code != 0) {
+      let error = new Error(stringifyExitStatus(status));
+      error.name = `ExecError`;
+      throw error;
+    } else {
+      return status;
+    }
+  }
+
+  let childProcess = spawnProcess(`${command}.cmd`, options.arguments || [], {
+    // We lose exit information and events if this is detached in windows
+    // and it opens a window in windows+powershell.
+    detached: false,
+    // When windows shell is true, it runs with cmd.exe by default, but
+    // node has trouble with PATHEXT and exe. It can't run exe directly for example.
+    // `cross-spawn` handles running it with the shell in windows if needed.
+    // Neither mac nor linux need shell and we run it detached.
+    shell: false,
+    // With stdio as pipe, windows gets stuck where neither the child nor the
+    // parent wants to close the stream, so we call it ourselves in the exit event.
+    stdio: "pipe",
+    env: options.env,
+    cwd: options.cwd,
+  });
+
+  let { pid } = childProcess;
+
+  return yield resource(
+    { pid, stdin, stdout, stderr, join, expect },
+    function*() {
+      let onError = (value: unknown) =>
+        getResult.resolve({ type: "error", value });
+
+      try {
+        childProcess.on("error", onError);
+
+        yield spawn(
+          on<[string]>(childProcess.stdout, "data").forEach(function*([data]) {
+            addToTail(data);
+            stdout.send(data);
+          })
+        );
+
+        yield spawn(
+          on<[string]>(childProcess.stderr, "data").forEach(function*([data]) {
+            addToTail(data);
+            stderr.send(data);
+          })
+        );
+
+        yield spawn(
+          subscribe(stdin).forEach(function*(data) {
+            childProcess.stdin.write(data);
+          })
+        );
+
+        let value = yield once(childProcess, "exit");
+        getResult.resolve({ type: "status", value });
+      } finally {
+        stdout.close();
+        stderr.close();
+        childProcess.off("error", onError);
+        try {
+          // @ts-ignore
+          childProcess.stdout.end();
+          // @ts-ignore
+          childProcess.stderr.end();
+          exec("taskkill /pid " + childProcess.pid + " /T");
```

Comment:
If I understand correctly, `exec()` will return a promise here, and so we need to call `.catch()` on it so that in the event that it fails, we don't get a `UhandledPromiseRejection` warning.

Reply:
`exec` takes a callback and returns a `<ChildProcess>`. We could use the promisified version, but I don't know that it would gain us anything.

> If this method is invoked as its util.promisify()ed version, it returns a Promise for an Object with stdout and stderr properties. The returned ChildProcess instance is attached to the Promise as a child property. In case of an error (including any error resulting in an exit code other than 0), a rejected promise is returned, with the same error object given in the callback, but with two additional properties stdout and stderr.

Reply:
Cool. Just wanted to make sure we didn't generate a bunch of spam to `stderr`